workflows/metatdenovo.nf

```diff
@@ -73,17 +153,378 @@ workflow METATDENOVO {
     // SUBWORKFLOW: Read in samplesheet, validate and stage input files
     //
     INPUT_CHECK (
-        ch_input
+        file(params.input)
     )
+    .reads
+    .map {
+        meta, fastq ->
+            new_id = meta.id - ~/_T\d+/
+            [ meta + [id: new_id], fastq ]
+    }
+    .groupTuple()
+    .branch {
+        meta, fastq ->
+            single  : fastq.size() == 1
+                return [ meta, fastq.flatten() ]
+            multiple: fastq.size() > 1
+                return [ meta, fastq.flatten() ]
+    }
+    .set { ch_fastq }
     ch_versions = ch_versions.mix(INPUT_CHECK.out.versions)
+    // TODO: OPTIONAL, you can use nf-validation plugin to create an input channel from the samplesheet with Channel.fromSamplesheet("input")
+    // See the documentation https://nextflow-io.github.io/nf-validation/samplesheets/fromSamplesheet/
+    // ! There is currently no tooling to help you write a sample sheet schema
 
     //
-    // MODULE: Run FastQC
+    // MODULE: Concatenate FastQ files from same sample if required
     //
-    FASTQC (
-        INPUT_CHECK.out.reads
+    CAT_FASTQ (
+        ch_fastq.multiple
     )
-    ch_versions = ch_versions.mix(FASTQC.out.versions.first())
+    .reads
+    .mix(ch_fastq.single)
+    .set { ch_cat_fastq }
+
+    ch_versions = ch_versions.mix(CAT_FASTQ.out.versions.first().ifEmpty(null))
+
+    //
+    // SUBWORKFLOW: Read QC and trim adapters
+    //
+    FASTQC_TRIMGALORE (
+        ch_cat_fastq,
+        params.skip_fastqc || params.skip_qc,
+        params.skip_trimming
+    )
+    ch_versions = ch_versions.mix(FASTQC_TRIMGALORE.out.versions)
+
+    FASTQC_TRIMGALORE.out.trim_log.collect { it[1] }
+    ch_collect_stats = ch_cat_fastq.collect { it[0].id }.map { [ [ id:"${params.assembler}.${params.orf_caller}" ], it ] }
+    if ( params.skip_trimming ) {
+        ch_collect_stats
+            .map { [ it[0], it[1], [] ] }
+            .set { ch_collect_stats }
+    } else {
+        if ( params.se_reads ) {
+            ch_collect_stats
+                .combine(FASTQC_TRIMGALORE.out.trim_log.collect { it[1] }.map { [ it ] })
+                .set { ch_collect_stats }
+        } else {
+            ch_collect_stats
+                .combine(FASTQC_TRIMGALORE.out.trim_log.collect { it[1][0] }.map { [ it ] })
+                .set { ch_collect_stats }
+        }
+    }
+
+    //
+    // MODULE: Run BBDuk to clean out whatever sequences the user supplied via params.sequence_filter
+    //
+    if ( params.sequence_filter ) {
+        BBMAP_BBDUK ( FASTQC_TRIMGALORE.out.reads, params.sequence_filter )
+        ch_clean_reads  = BBMAP_BBDUK.out.reads
+        ch_bbduk_logs = BBMAP_BBDUK.out.log.collect { it[1] }.map { [ it ] }
+        ch_versions   = ch_versions.mix(BBMAP_BBDUK.out.versions)
+        ch_collect_stats
+            .combine(ch_bbduk_logs)
+            .set {ch_collect_stats}
+        ch_multiqc_files = ch_multiqc_files.mix(BBMAP_BBDUK.out.log.collect{it[1]}.ifEmpty([]))
```

Comment:
To avoid the presence of multiple `[]` in channels, only use `ifEmpty` in the final process call to multiqc. This cuts down on code.